File: ok2curl/src/test/java/com/moczul/ok2curl/LimitedSinkTest.java
package com.moczul.ok2curl;

import org.junit.Test;

import java.nio.charset.Charset;

import okio.Buffer;
import okio.BufferedSink;
import okio.ByteString;
import okio.Okio;

import static org.junit.Assert.*;

public class LimitedSinkTest {

    @Test(expected = IllegalArgumentException.class)
    public void testWhenLimitIs0_throwIllegalArgumentException() throws Exception {
        new LimitedSink(new Buffer(), 0L);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testWhenLimitIsNegative_throwIllegalArgumentException() throws Exception {
        new LimitedSink(new Buffer(), -1L);
    }

    @Test(expected = NullPointerException.class)
    public void testWhenBufferIsNull_throwNullPointerException() throws Exception {
        new LimitedSink(null, 10L);
    }

    @Test
    public void testWhenLongResult_writeOnlyLimitedBytes() throws Exception {
        final Buffer limited = new Buffer();
        final BufferedSink source = Okio.buffer(new LimitedSink(limited, 10L));

        source
                .write(ByteString.encodeUtf8("0123456789012345678901234567890123456789")); // 40B
        source.flush();

        assertEquals("0123456789", limited.readString(Charset.forName("UTF-8")));
    }

    @Test
    public void testWhenLongResultWrittenInParts_writeOnlyLimitedBytes() throws Exception {
        final Buffer limited = new Buffer();
        final BufferedSink source = Okio.buffer(new LimitedSink(limited, 10L));

        source
                .write(ByteString.encodeUtf8("01234")); // 5B
        source.flush();
        source
                .write(ByteString.encodeUtf8("5678901234")); // 10B
        source.flush();
        source
                .write(ByteString.encodeUtf8("5678901234567890123456789")); // 25B
        source.flush();

        assertEquals("0123456789", limited.readString(Charset.forName("UTF-8")));
    }

    @Test
    public void testWhenShorterResult_writeOnlyThatData() throws Exception {
        final Buffer limited = new Buffer();
        final BufferedSink source = Okio.buffer(new LimitedSink(limited, 10L));

        source
                .write(ByteString.encodeUtf8("01234")); // 5B
        source.flush();

        assertEquals("01234", limited.readString(Charset.forName("UTF-8")));
    }
}

File: ok2curl/src/test/java/com/moczul/ok2curl/LimitedSinkTest.kt
package com.moczul.ok2curl

import okio.Buffer
import okio.ByteString.Companion.encodeUtf8
import okio.buffer
import org.junit.Assert.assertEquals
import org.junit.Test
import java.nio.charset.Charset

class LimitedSinkTest {

    @Test(expected = IllegalArgumentException::class)
    fun `should throw exception when limit is 0`() {
        LimitedSink(Buffer(), 0L)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `should throw exception when limit is negative`() {
        LimitedSink(Buffer(), -1L)
    }

    @Test(expected = NullPointerException::class)
    fun `should throw exception when buffer is null`() {
        LimitedSink(null, 10L)
    }

    @Test
    fun `should write to buffer only limited amount of bytes`() {
        // given
        val limited = Buffer()
        val source = LimitedSink(limited, 10L).buffer()

        // when
        source.write("0123456789012345678901234567890123456789".encodeUtf8()) // 40B
        source.flush()

        // then
        assertEquals("0123456789", limited.readString(Charset.forName("UTF-8")))
    }

    @Test
    fun `should write to buffer only limited amount of bytes when written in parts`() {
        // given
        val limited = Buffer()
        val source = LimitedSink(limited, 10L).buffer()

        // when
        source.write("01234".encodeUtf8()) // 5B
        source.flush()
        source.write("5678901234".encodeUtf8()) // 10B
        source.flush()
        source.write("5678901234567890123456789".encodeUtf8()) // 25B
        source.flush()

        // then
        assertEquals("0123456789", limited.readString(Charset.forName("UTF-8")))
    }

    @Test
    fun `should write all bytes when limit is bigger than input`() {
        // given
        val limited = Buffer()
        val source = LimitedSink(limited, 10L).buffer()

        // when
        source.write("01234".encodeUtf8()) // 5B
        source.flush()

        // then
        assertEquals("01234", limited.readString(Charset.forName("UTF-8")))
    }
}